hwdetect/hwdetect.sh

```diff
@@ -0,0 +1,319 @@
+    local major=$(get_device_major ${device})
+    local minor=$(get_device_minor ${device})
+    local model=$(get_usb_to_ttl_convertor_model ${device})
+
+    printf "%-15s %-10s %-s\n" "${device}" "${major}:${minor}" "${model}"
+  done
+
+  echo
+}
+
+print_hardware_listing() {
+  print_block_devices "$(get_block_devices)"
+  print_i2c_devices "$(get_i2c_devices)"
+  print_usb_to_ttl_convertors "$(get_usb_to_ttl_convertors)"
+}
+
+################################################################################
+
+readonly DEFAULT_HARDWARE_DETECTION_DELAY=10
+readonly HARDWARE_BEFORE_SNAPSHOT="/tmp/hwdetect-before.snapshot"
+readonly HARDWARE_AFTER_SNAPSHOT="/tmp/hwdtect-after.snapshot"
+
+write_devices_to_hardware_snapshot() {
+  local readonly snapshot=$1
+  local readonly label=$2
+  local readonly devices=($3)
+
+  for device in "${devices[@]}"; do
+    printf "%s#%s\n" "${label}" "${device}" >> ${snapshot}
+  done
+}
+
+create_hardware_snapshot() {
+  local readonly snapshot=$1
+
+  printf "timestamp %s\n" "$(date +'%Y%m%d-%H:%M:%S.%N')" > ${snapshot}
+
+  write_devices_to_hardware_snapshot ${snapshot} "blk" "$(get_block_devices)"
+  write_devices_to_hardware_snapshot ${snapshot} "i2c" "$(get_i2c_devices)"
+  write_devices_to_hardware_snapshot ${snapshot} "ttyUSB" \
+    "$(get_usb_to_ttl_convertors)"
+}
+
+watch_hardware_changes() {
+  local readonly hardware_detection_delay=$1
+
+  printf "HARDWARE CHANGES\n"
+  printf "================\n"
+  printf "%-28s %-20s %-10s %s\n" "TIMESTAMP" "DEVICE" "TYPE" "STATUS"
+
+  while true; do
+    create_hardware_snapshot ${HARDWARE_BEFORE_SNAPSHOT}
+    sleep ${hardware_detection_delay}
+    create_hardware_snapshot ${HARDWARE_AFTER_SNAPSHOT}
+
+    local hardware_diff=$(
+      diff \
+      <(cat ${HARDWARE_BEFORE_SNAPSHOT} | tail -n +2) \
+      <(cat ${HARDWARE_AFTER_SNAPSHOT} | tail -n +2) \
+      | tail -n +2
+    )
+
+    if [[ -z "${hardware_diff}" ]]; then
+      continue
+    fi
+
+    local readonly timestamp=$(date +"%Y%m%d-%H:%M:%S.%N")
+
+    local readonly connected_hardware=(
+      $(echo "${hardware_diff}" | grep ">" | cut -d " " -f2)
+    )
+    local readonly disconnected_hardware=(
+      $(echo "${hardware_diff}" | grep "<" | cut -d " " -f2)
+    )
+
+    for hardware in "${connected_hardware[@]}"; do
+      local device=$(echo ${hardware} | cut -d "#" -f2)
+      local type=$(echo ${hardware} | cut -d "#" -f1)
+
+      printf "%-28s %-20s %-10s %s\n" "${timestamp}" "${device}" "${type}" \
+        "CONNECTED"
+    done
+
+    for hardware in "${disconnected_hardware[@]}"; do
+      local device=$(echo ${hardware} | cut -d "#" -f2)
+      local type=$(echo ${hardware} | cut -d "#" -f1)
+
+      printf "%-28s %-20s %-10s %s\n" "${timestamp}" "${device}" "${type}" \
+        "DISCONNECTED"
+    done
+  done
+}
+
+################################################################################
+
+readonly OPTIONS="lwd:h"
+readonly OPTERR=0
+
+main() {
```

Comment:
Is it required to create function main? It is just a question. May be Yes for your testing purposes?

Reply:
https://google.github.io/styleguide/shell.xml?showone=main#main